File: array/0081_search_in_rotated_sorted_array_ii/0081_search_in_rotated_sorted_array_ii.py
class Solution(object):
    def search(self, nums, target):
        left, right = 0, len(nums) - 1

        while left <= right:
            mid = left + (right - left) / 2

            if nums[mid] == target:
                return True
            elif nums[mid] == nums[left]:
                left += 1
            elif (nums[mid] > nums[left] and nums[left] <= target < nums[mid]) or (nums[mid] < nums[left] and not (nums[mid] < target <= nums[right])):
                right = mid - 1
            else:
                left = mid + 1

        return False

nums = [2,5,6,0,0,1,2]
target = 0 

res = Solution().search(nums, target)
print(res)

File: array/0081_search_in_rotated_sorted_array_ii/0081_search_in_rotated_sorted_array_ii.go
package main 

import "fmt"

func search(nums []int, target int) bool {
	length := len(nums)

	if length == 0 {
		return false
	}

	k:= 1
	for k < len(nums) && nums[k-1] <= nums[k] {
		k++
	}
	fmt.Println("k is", k)
	i, j := 0, length -1
	for i <= j{
		m := (i + j)/ 2
		med := (m+k) % length
		fmt.Println("m is", m, "med is", med)

		switch {
		case nums[med] < target:
			i = m + 1
		case target < nums[med]:
			j = m - 1
		default:
			return true
		}
	}

	return false
}

func main() {
	nums := []int{2,5,6,0,0,1,2}
	target := 0
	res := search(nums, target)
	fmt.Println(res)
}